relay-server/src/services/server.rs

```diff
@@ -55,115 +40,204 @@ const CLIENT_HEADER_TIMEOUT: Duration = Duration::from_secs(5);
-            .layer(middleware)
-            .with_state(service);
-
-        // Bundle middlewares that need to run _before_ routing, which need to wrap the router.
-        // ConnectInfo is special as it needs to last.
-        let app = ServiceBuilder::new()
-            .layer(NormalizePathLayer::new())
-            .service(router)
-            .into_make_service_with_connect_info::<SocketAddr>();
-
-        let http_config = HttpConfig::new()
-            .http1_half_close(true)
-            .http1_header_read_timeout(CLIENT_HEADER_TIMEOUT)
-            .http1_writev(true)
-            .http2_keep_alive_timeout(config.keepalive_timeout())
-            .build();
-
-        let addr_config = AddrIncomingConfig::new()
-            .tcp_keepalive(Some(config.keepalive_timeout()).filter(|d| !d.is_zero()))
-            .tcp_keepalive_interval(Some(config.keepalive_timeout()).filter(|d| !d.is_zero()))
-            .tcp_keepalive_retries(Some(KEEPALIVE_RETRIES))
-            .build();
-
-        let handle = Handle::new();
-
-        match create_listener(config.listen_addr(), config.tcp_listen_backlog()) {
-            Ok(listener) => {
-                listener.set_nonblocking(true).ok();
-                let server = axum_server::from_tcp(listener)
-                    .http_config(http_config)
-                    .addr_incoming_config(addr_config)
-                    .handle(handle.clone());
-
-                relay_log::info!("spawning http server");
-                relay_log::info!("  listening on http://{}/", config.listen_addr());
-                relay_statsd::metric!(counter(RelayCounters::ServerStarting) += 1);
-                tokio::spawn(server.serve(app));
-            }
-            Err(err) => {
-                relay_log::error!("Failed to start the HTTP server: {err}");
-                std::process::exit(1);
+    let mut ka = TcpKeepalive::new().with_time(ka_timeout);
+    #[cfg(not(any(target_os = "openbsd", target_os = "redox", target_os = "solaris")))]
+    {
+        ka = ka.with_interval(ka_timeout);
+    }
+
+    #[cfg(not(any(
+        target_os = "openbsd",
+        target_os = "redox",
+        target_os = "solaris",
+        target_os = "windows"
+    )))]
+    {
+        ka = ka.with_retries(KEEPALIVE_RETRIES);
+    }
+
+    Some(ka)
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct KeepAliveAcceptor(Option<TcpKeepalive>);
+
+impl KeepAliveAcceptor {
+    /// Create a new acceptor that sets TCP_NODELAY and keep-alive.
+    pub fn new(config: &Config) -> Self {
+        Self(build_keepalive(config))
+    }
+}
+
+impl<S> Accept<TcpStream, S> for KeepAliveAcceptor {
+    type Stream = TcpStream;
+    type Service = S;
+    type Future = std::future::Ready<io::Result<(Self::Stream, Self::Service)>>;
+
+    fn accept(&self, stream: TcpStream, service: S) -> Self::Future {
+        if let Self(Some(ref tcp_keepalive)) = self {
+            let sock_ref = socket2::SockRef::from(&stream);
```

Comment:
possibly an option, if you build the socket via Tokio's `TcpSocket` you can set the keepalive directly on the [socket](https://docs.rs/tokio/1.39.3/tokio/net/struct.TcpSocket.html#method.set_keepalive).

Reply:
We can set the keepalive flag, but not configure the keepalive interval and other options. I could keep the SockRef and move it to where we bind the initial socket, though. 

Looking hyper and other implementors, they for some reason always do this on the _accepted_ TCP stream rather than the listen address. I wasn't sure whether that made a difference, so I chose to keep it that way.